### PR TITLE
fix: reorder AST folding after semantics validation

### LINKED ISSUE #2830: Type of integer constant is ignored during folding of literal ops
### Version Information

* vyper Version (output of `vyper --version`): 0.3.3
* OS: Linux
* Python Version (output of `python --version`): 3.8.5

### What's your issue about?

Please include information like:

This contract compiles when it should throw a typechecking error.
```
a: constant(uint16) = 200

@external
def foo() -> int16:
	return a - 201
```

### How can it be fixed?

Annotate the folded nodes with the original type information.

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix #2830, and stricter type safety for literal operations.

### How I did it

- Reorder AST folding to after semantics validation.
- Added `not_assignable` flag to `BaseTypeDefinition`.
- Added annotation of constants, and helper functions to derive foldable values where required for semantics validation before AST nodes are folded.

### How to verify it

See new tests, and updated tests.

### Commit message

```
fix: reorder AST folding after semantics validation

Fixes an issue with type-checking and bounds-check of literal ops.

With the reordering, a `not_assignable` flag has been added to 
`BaseTypeDefinition`, aligning with the intent behind `check_kwargable`.
This allows true constants (builtin or user-defined) to be differentiated
from calldata arguments which are not assignable, but not constants.

Helper functions have also been added to support the derivation of foldable
values for semantics validation and to determine whether a node is foldable
 prior to their AST folding.

Fix #2830
```
### Description for the changelog

Type-check and bounds-check literal ops involving literals from constant folding.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ast/nodes.py [hunk 1]
```diff
@@ -1086,6 +1106,36 @@ def evaluate(self) -> VyperNode:
...
+            raise UnfoldableNode(
+                f"Cannot perform {self.op._description} comparison between dislike types"
+            )
+
+        if None in types and len(types) == 2:
+            if left_type and not SizeLimits.in_bounds(repr(left_type), right.value):
+                # Check right value is in bounds
+                raise OverflowException(
+                    (
+                        f"The right literal's ({right.value}) is outside bounds of the ",
```
Comment (charles-cooper) [inline vyper/ast/nodes.py:1124]: i think we can assume in general that users are not super familiar with compiler jargon ("literal", etc.). what if we printed something like `"{right.value} cannot be {description} with {left.value} (which is of type {left_type})"`?

**Review by charles-cooper — COMMENTED** [review]
looks like there is some code that is duplicated in structure. is there some way we can refactor `_validate_numeric_bounds` to address these cases?

**Comment by charles-cooper** [comment]
it is also very weird to see typechecking code in AST construction. is it possible to move these to the constant folder module?

**Comment by tserg** [comment]
> looks like there is some code that is duplicated in structure. is there some way we can refactor `_validate_numeric_bounds` to address these cases?

I have moved the typechecking to `folding.py` so it only runs after `_validate_numeric_bounds` for `BinOp` and `Compare` nodes.

**Comment by tserg** [comment]
Note: We may want to add builtin constants for `MAX_UINT`, `MIN_INT` and `MAX_INT` for all integer types after this change. Otherwise, it is difficult to assign the min/max value in an easy way because the exponentiation value will be out of bounds.

**Comment by tserg** [comment]
Seems like there is regression in bytecode size - e.g.
```
@external
def foo() -> uint256:
    a: uint256 = 2 ** 200
    b: uint256 = 5
    return max(a, 5) + max(b, 5)
```
```
@external
def goo() -> uint256:
    return min(3, 5) + max(40, 80)
```

**Comment by tserg** [comment]
> Note: We may want to add builtin constants for `MAX_UINT`, `MIN_INT` and `MAX_INT` for all integer types after this change. Otherwise, it is difficult to assign the min/max value in an easy way because the exponentiation value will be out of bounds.

Pending #2935. Tests for bounds should be replaced with this builtin once merged.

**Pull request closed without merging**